src/whylogs/core/statistics/constraints.py

```diff
@@ -0,0 +1,272 @@
+    Op.LE: lambda f, f2: lambda s: getattr(s, f) <= getattr(s, f2),
+    Op.EQ: lambda f, f2: lambda s: getattr(s, f) == getattr(s, f2),
+    Op.NE: lambda f, f2: lambda s: getattr(s, f) != getattr(s, f2),
+    Op.GE: lambda f, f2: lambda s: getattr(s, f) >= getattr(s, f2),
+    Op.GT: lambda f, f2: lambda s: getattr(s, f) > getattr(s, f2),
+}
+
+
+class ValueConstraint:
+    """
+    ValueConstraints express a binary boolean relationship between an implied numeric value and a literal.
+    These are applied to every incoming value that is processed by whylogs.
+    """
+
+    def __init__(self, op: Op, value, name: str = None, verbose=False):
+        self._name = name
+        self._verbose = verbose
+        self.op = op
+        self.value = value
+        self.func = _value_funcs[op](value)
+        self.total = 0
+        self.failures = 0
+
+    @property
+    def name(self):
+        return self._name if self._name is not None else f'value {Op.Name(self.op)} {self.value}'
+
+    def track(self, v) -> bool:
+        self.total += 1
+        if not self.func(v):
+            self.failures += 1
+            if self._verbose:
+                logger.info(f'value constraint {self.name} failed on value {v}')
+
+    @staticmethod
+    def from_protobuf(msg: ValueConstraintMsg) -> 'ValueConstraint':
+        return ValueConstraint(msg.op, msg.value, name=msg.name)
+
+    def to_protobuf(self) -> ValueConstraintMsg:
+        return ValueConstraintMsg(
+            name=self.name,
+            op=self.op,
+            value=self.value,
+        )
+
+    def report(self):
+        return (self.name, self.total, self.failures)
+
+
+class SummaryConstraint:
+    """
+    Summary constraints specify a relationship between a summary field and a literal value,
+    or between two summary fields.
+    e.g.     'min' < 6
+             'std_dev' < 2.17
+             'min' > 'avg'
+    """
+
+    def __init__(self, field: str, op: Op, value=None, field2: str = None, name: str = None, verbose=False):
+        self._verbose = verbose
+        self._name = name
+        self.op = op
+        self.field = field
+        self.field2 = field2
+        self.total = 0
+        self.failures = 0
+        if value is not None and field2 is None:
+            # field-value summary comparison
+            self.value = value
+            self.func = _summary_funcs1[op](field, value)
+        elif field2 is not None and value is None:
+            # field-field summary comparison
+            self.field2 = field2
+            self.func = _summary_funcs2[op](field, field2)
+        else:
+            raise ValueError("Summary constraint must specify a second value or field name, but not both")
+
+    @property
+    def name(self):
+        return self._name if self._name is not None else f'summary {self.field} {Op.Name(self.op)} {self.value}/{self.field2}'
+
+    def track(self, summ: NumberSummary) -> bool:
```

Comment:
docstring?

We use `track` when tracking datapoints. So maybe use a different name here? `update` for example

Reply:
👍